packages/babel-plugin-transform-react-display-name/src/index.js

```diff
@@ -42,53 +69,161 @@ export default function ({ types: t }) {
+          let displayName;
+          if (path.parentPath.node.type === "ExportDefaultDeclaration") {
+            if (path.node.id == null) {
+              // An anonymous function declaration in export default declaration.
+              // Transform `export default function () { ... }`
+              // to `var _uid1 = function () { .. }; export default __uid;`
+              // then add displayName to _uid1
+              const extension = pathMod.extname(state.file.opts.filename);
+              const name = pathMod.basename(state.file.opts.filename, extension);
+
+              const id = path.scope.generateUidIdentifier("uid");
+              path.node.id = id;
+              displayName = name;
+            }
+            setDisplayNameAfter(path, path.node.id, displayName);
+          } else if (path.parentPath.node.type === "Program"
+                     || path.parentPath.node.type == "ExportNamedDeclaration") {
+            setDisplayNameAfter(path, path.node.id, displayName);
+          }
+        }
+      },
+      FunctionExpression: function (path) {
+        if (shouldSetDisplayNameForFuncExpr(path)) {
+          const id = findCandidateNameForExpression(path);
+          if (id) {
+            setDisplayNameAfter(path, id);
+          }
+        }
+      },
+      ArrowFunctionExpression: function (path) {
+        if (shouldSetDisplayNameForFuncExpr(path)) {
+          const id = findCandidateNameForExpression(path);
+          if (id) {
+            setDisplayNameAfter(path, id);
+          }
         }
       },
     },
   };
 }
+
+function shouldSetDisplayNameForFuncExpr(path) {
+  // Parent must be either 'AssignmentExpression' or
+  // 'VariableDeclarator' or 'CallExpression' with a parent of 'VariableDeclarator'
+  if (path.parentPath.node.type === "AssignmentExpression" &&
+      path.parentPath.node.left.type !== "MemberExpression" && // skip static members
+      path.parentPath.parentPath.node.type == "ExpressionStatement" &&
+      path.parentPath.parentPath.parentPath.node.type == "Program") {
+    return doesReturnJSX(path.node.body);
+  } else {
+    // if parent is a call expression, we have something like (function () { .. })()
+    // move up, past the call expression and run the rest of the checks as usual
+    if (path.parentPath.node.type === "CallExpression") {
+      path = path.parentPath;
+    }
+
+    if (path.parentPath.node.type === "VariableDeclarator") {
+      if (path.parentPath.parentPath.parentPath.node.type === "ExportNamedDeclaration" ||
+          path.parentPath.parentPath.parentPath.node.type === "Program") {
+        return doesReturnJSX(path.node.body);
+      }
+    }
+  }
+
+  return false;
+}
+
+function classHasRenderMethod(node) {
+  if (!node.node.body) {
+    return false;
+  }
+  const members = node.node.body.body;
+  for (let i = 0; i < members.length; i++) {
+    if (members[i].type == "ClassMethod" && members[i].key.name == "render") {
+      return true;
```

Comment:
Checking if the class has a non-`null` `extends` clause could help reduce false positives, as well as checking if `render` `doesReturnJSX`.